### PR TITLE
[SPARK-45358][SQL] Remove shim classes for Hive prior 2.0.0

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

This PR aims to clean up unused Hive shim classes for Hive prior 2.0.0, it's pure refactor, kind of something like `git squash`

### Why are the changes needed?

SPARK-45328 removed Hive support prior to 2.0.0, this PR is the next step of the code clean up, see https://github.com/apache/spark/pull/43116#discussion_r1336612849

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass GA

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by pan3793** [comment]
cc @HyukjinKwon @LuciferYang

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 1]
```diff
@@ -255,11 +254,19 @@ private[client] sealed abstract class Shim {
   }
 }
 
-private class Shim_v0_12 extends Shim with Logging {
+private[client] class Shim_v2_0 extends Shim with Logging {
   // See HIVE-12224, HOLD_DDLTIME was broken as soon as it landed
   protected lazy val holdDDLTime = JBoolean.FALSE
```
Comment (LuciferYang) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:259]: `holdDDLTime` become a useless val?

Comment (pan3793) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: right, removed

**Comment by LuciferYang** [comment]
https://github.com/apache/spark/blob/2cc1ee4d3a05a641d7a245f015ef824d8f7bae8b/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala#L264-L269

Can `getMSCMethod` use `findMethod()` now?

**Comment by pan3793** [comment]
> Can `getMSCMethod` use `findMethod()` now?

I think no, `findMethod` only works for public method.
```
  protected def findMethod(klass: Class[_], name: String, args: Class[_]*): Method = {
    klass.getMethod(name, args: _*)
  }
```

BTW, to make the change clear, the change in this PR is conservative, in the next step, we can replace some method reflection with direct API call.

**Comment by LuciferYang** [comment]
Next, we might consider using `MethodHandle` or `VarHandle` instead of reflection, they should be the more recommended way to use in Java 17+.

**Comment by LuciferYang** [comment]
The code changes are fine to me, hoping to get more reviews from others.

**Review by HyukjinKwon — COMMENTED** [review]
I'm fine with this. Looks good from a cursory look

**Comment by HyukjinKwon** [comment]
cc @dongjoon-hyun

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

cc @sunchao , too

**Review by sunchao — APPROVED** [review]
LGTM too

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**